### ISSUE TITLE
Magic remote/Mouse support

### ISSUE THREAD

**Issue by BlueBazze** [spacerefugee/vue-js-spatial-navigation#4]
So far adding:
```
mounted(el, binding) {
        el.addEventListener("mouseenter", function () {
          Vue.config.globalProperties.$SpatialNavigation.focus(el);
        });
      },
```
To the v-focus directive. Adds the ability to focus on elements with the mouse.
We might be able to have a better way later on.

For the on click trigger.
Previously we used the "enter-down" as defined here: [https://github.com/spacerefugee/vue-js-spatial-navigation/blob/master/lib/spatial_navigation.js#L879-L889](https://github.com/spacerefugee/vue-js-spatial-navigation/blob/master/lib/spatial_navigation.js#L879-L889)

We could change the event to trigger the regular click event instead of the custom. This could be a breaking change to existing projects.

We want to trigger currentFocusedElement click instead of
```
if (!fireEvent(currentFocusedElement, "enter-down")) {
    return preventDefault();
}
```

The end result when defining buttons would look like
Vue2
```
<v-btn v-focus v-on:click="myFunction">Click me</v-btn>
```

Vue3
```
<v-btn v-focus @click="myFunction">Click me</v-btn>
```

Both clicking with a mouse & pressing enter on that element would trigger the @click event.

**Comment by BlueBazze** [comment]
So far the only way i have gotten these two features to work is with the v-focus directive.
Which looks like this
```
Vue.directive("focus", {
      beforeMount(el, binding) {
        disableElement(el, binding.value);
      },
      mounted(el, binding) {
        el.addEventListener("mouseenter", function () {
          Vue.config.globalProperties.$SpatialNavigation.focus(el);
        });
        el.addEventListener("keyup", function (e) {
          if (e.keyCode === 13) {
            el.click();
          }
        });
      },
      updated(el, binding) {
        disableElement(el, binding.value);
      },
      unmounted(el) {
        el.removeAttribute("data-focusable");
      },
    });
```

I tried to implement the click directly into spatial_navigation.js
Instead of this https://github.com/spacerefugee/vue-js-spatial-navigation/blob/master/lib/spatial_navigation.js#L879-L889

i tried
```
if (!direction) {
    if (evt.keyCode == 13) {
      currentFocusedElement = getCurrentFocusedElement();
      if (currentFocusedElement && getSectionId(currentFocusedElement)) {
        currentFocusedElement.click();
      }
    }
    return;
  }
```
Which does absolutely nothing.


I am still very new to web development. So there might very well be a better way to handle these two features, and a better way to integrate it with vue. Ill keep looking for it, but i havent found it yet.

**Comment by spacerefugee** [comment]
Maybe you can trigger the keydown event manually when clicking the focused element.
Try to add this code in `v-focus`:
```javascript
 Vue.directive("focus", {
    ...
    mounted(el, binding) {
      el.addEventListener("mouseenter", function () {
        SpatialNavigation.focus(el);
      });
      el.addEventListener("click", function (e) {
        el.dispatchEvent(new KeyboardEvent("keydown", { keyCode: 13 }));
      });
    }
    ...
  });
````

**Comment by BlueBazze** [comment]
Yes, i do see your point with this. The only reason i did it the other way around. Is simply the usability for the developer. If a new guys comes in, the {@}click might be easier to understand, with enter key just triggering that.

Regarding the mouse enter event. I thought i tried this instead of going through global properties / prototype before. But it works fine now.

Tho one thing i have noticed with keydown/up. You kinda want navigation on keydown & actions on keyup. Because keydown is fires multiple times if you hold it down. While keyup only fires when the key is released.

If you take a look at netflix, when navigating horizontally to the right. Its not loading the information on the top half of the screen, for every movie you navigate over, but only on the one you stop at.
Either you would have to debounce all those calls to get the info about that movie, or you could use a keyup listener on every movie.

I use a combo of both keydown & keyup. Mostly all actions are controlled by keyup. While only a few actions like a date picker is controlled by keydown for continously changing the date.

Using keydown with enter, could lead to some users experiencing issues, especially elders. If they hold down the enter key. The event could fire multiple times.

**Comment by BlueBazze** [comment]
Trying to unify everything. I have added a new function to spatial_navigation.js
```js
function onClick(event) {
  var currentFocusedElement = getCurrentFocusedElement();
  if (
    event.target == currentFocusedElement &&
    getSectionId(currentFocusedElement)
  ) {
    if (!fireEvent(currentFocusedElement, "enter-up")) {
      evt.preventDefault();
      evt.stopPropagation();
    }
  }
}
```
In the init function:
`window.addEventListener("click", onClick);`
In the uninit function:
`window.removeEventListener("click", onClick);`

Making it so we wont have to have a billion event listeners, instead just one checking if we have clicked the focused element. We can then move the focus around with mouse enter eventlisteners.

**Comment by BlueBazze** [comment]
Ive written a short example of this at https://github.com/Syncronet-APS/vue-js-spatial-navigation/blob/master/docs/v-focus.md#button-click

The enter-up will execute on both enter press & mouse click.

The commit for the update in spatial_navigation.js can be found here https://github.com/Syncronet-APS/vue-js-spatial-navigation/commit/65f508be3df82fc0e9008320bd40e3b8e9d7197b

**Comment by BlueBazze** [comment]
I have been thinking if it would be worth the cost, adding a mouseenter event listener to the window. And checking if its a focusable element. Instead of having the directive event listener. I dont see any benefits other than being unified in spatial_navigation.js but the major downside is that it would trigger on every element that is entered with the mouse and then checking if its focusable.

**Comment by spacerefugee** [comment]
If you are just trying to make the mouse click event behavior like `keydown` and `keyup` events, the `mousedown` and `mouseup` events should be work. 
For the event listener should binding on window or every elements, I think if it doesn't cost too much performance you can just use what you like to achieve this.

**Comment by BlueBazze** [comment]
Short status. I have everything working as expected. There are tho a few missing things. So far i havent been bothered with making a directive for click-outside of an element. It comes with Vuetify. And they are aiming for a beta release in feburary. This is relevant for the product i am working on, so it might be relevant for others.

Other than that. The commits ive made on that fork should somewhat be compatible with v2. Keep in mind that ive updated it to v3. Later when i get the time for it, i will take a look through the commits on other js-spatial-navigation forks that ive listed in my fork for v3.
